src/python/WMComponent/AgentStatusWatcher/DrainStatusPoller.py

```diff
@@ -73,6 +78,39 @@ def getDrainInfo(cls):
         """
         return cls.drainStats
 
+    def checkDrainStatusChanges(self):
```

Comment:
For this method, I have a couple of comments:
a) the agent might have had multiple drain transitions, and in this case we would alert only on the last one evaluated.
b) I think we can reduce this code by iterating over the drain keys and using self.getattr(blah.lower()) or sort of to evaluate things in a loop and send an email alert from inside the loop (such that we catch multiple cases).

The message could even be: `"Agent had a drain status transition. %s = %s" % key, value` or something like that.

Reply:
At first I thought we would need to handle multiple transitions, for instance user drain and speed drain. But, these can only be caught if we make another call to aux db somewhere in the DrainStatusPoller algorithm or in checkDrainStatusChanges, which I was trying to avoid.

For example:
1. User sets aux db to UserDrainMode=True when jobs are < Speed drain thresholds
2. DrainStatusPoller algorithm will run and query aux db caching current drain statuses
3. Agent enters speed drain mode and updates aux db

In this case the agent has an out of date version of aux db contents (self.agentConfig) and we can't make the status comparison to alert on. Even if I run checkDrainStatusChanges at the end of the DrainStatusPoller algorithm function, it won't pick up the speed drain changes since my state comparisons used the out of date config.

If it is OK to make an additional call to the aux db in checkDrainStatusChanges, then doing multiple transitions in a single alert is definitely doable. I'll just move that function to the end of the DrainStatusPoller algorithm.

I guess there could be a rare case where the agent would enter both UserDrainMode and AgentDrainMode between DrainStatusPoller runs as well.

Reply:
I'm not sure I completely understood the case you pointed out, but let me try it from another angle.

We have the local object variable containing the drain statusES of the previous thread run. In the line 44 we fetch data from ReqMgrAux:
https://github.com/dmwm/WMCore/blob/master/src/python/WMComponent/AgentStatusWatcher/DrainStatusPoller.py#L44

which then we compare inside this function to see whether there were changes; if there were, create an alert.

Then the missing part is, along the algorithm() executing of this Drain thread, we need to trigger another email in case there are further drain transitions. And last but not least, also update the object variable for the next cycle.

IF we don't trigger an alert in other places where a transition can happen (based on thresholds and etc), then we'd receive an alert one cycle later.

Does it make sense?